admin/data/sql/migrations/20190918140848.do.pupil-state.sql

```diff
@@ -0,0 +1,21 @@
+alter table mtc_admin.pupil
+	add currentCheckId int
+
+alter table mtc_admin.pupil
+	add checkComplete bit
+	constraint DF_checkComplete_Default default (0)
+
+alter table mtc_admin.pupil
+	add restartActive bit
```

Comment:
Just reading this cold I am not sure whether `restartActive` indicates that permission for a restart has been granted, or whether a restart has been granted *and* the pin has been generated, or something else.

Of course, it's a good name, so perhaps I will learn more later on in the PR.

Reply:
Good point, i will add a readme to supplement the changes.

restartActive is populated when the restart is created.  This serves to indicate that the pupil can be included in the list of pupils eligible for pin generation.

prior to this, the following will happen...
- a restart record will be created in pupilRestart
- the restart id will be added to the check that is now void
- this `restartActive` flag will be set to true

Reply:
That's great 😄 

I wonder if I might suggest another name then...`restartPermission` or `permissionToRestart` ?  As it _seems_ to be a permission to generate another pin despite the pupil already having a check in an end-state.

If this was _code_ we would most likely prefix it with `has` or `is` to indicate that it is a boolean.

I also like the fact that with this system a DBA could grant permission for an extra restart in the case where all system-allowed restarts had been used, by following the above steps manually.  The pin generation code would show the pupil in the list just on the basis of the flag. 👏

Reply:
@GuyHarwood wouldn't it be better to add the restart id to the new check so that it is indicated that this check is a restart?

Reply:
then you have this dangling restart waiting for a new check.  I have never liked the existing design, as it requires looking up the restart when you generate a pin.  The restart is essentially just a token that grants a new check to be taken, and has more to do with the existing check being voided.

Reply:
@jon-shipley @GeorgeChatzigiannis how about `restartAvailable` instead of `restartActive` ?

Reply:
Personally I like `restartActive` as it points to the fact that this pupil has a restart that can be consumed/used.
Perhaps we could make it a bit more 'boolean-looking' by naming it hasActiveRestart but I still prefer `restartActive` over all the recommendations.

Reply:
i prefer `restartAvailable` as that is more indicative of something that is ready to use. 🙂